nifty-ssl/src/main/java/com/facebook/nifty/ssl/CryptoUtil.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2012-2013 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.nifty.ssl;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.BaseEncoding;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * Crypto-related utilities.
+ */
+public class CryptoUtil {
+
+    /** Length of a SHA256 hash, in bytes */
+    public static final int SHA256_OUTPUT_BYTES = 256 / 8;
+
+    /** Maximum value for the outputLength parameter to hkdf(). */
+    public static final int MAX_HKDF_OUTPUT_LENGTH = SHA256_OUTPUT_BYTES * 255;
+
+    /** Default all-0 salt used by HKDF if a null salt parameter is provided. */
+    private static final byte[] NULL_SALT = new byte[SHA256_OUTPUT_BYTES];
+
+    /** Empty byte array. */
+    private static final byte[] EMPTY_BYTES = new byte[0];
+
+    /** Name of the HMAC algorithm used by hkdf() function. */
+    private static final String MAC_ALGORITHM = "HmacSHA256";
+
+    /**
+     * Initializes a {@link Mac} object using the given key.
+     *
+     * @param key the HMAC key.
+     * @return the initialized Mac object.
+     * @throws IllegalArgumentException if the provided key is invalid.
+     */
+    private static Mac initHmacSha256(final byte[] key) {
+        try {
+            SecretKeySpec keySpec = new SecretKeySpec(key, MAC_ALGORITHM);
+            final Mac mac = Mac.getInstance(MAC_ALGORITHM);
```

Comment:
Definitely premature. The hkdf() function will only be called when we re-read the ticket seed file, which will happen every N minutes (with maybe N = 5 or so?). Or if we can register a file system listener that watches for file changes, it will only happen when the ticket seed file is rotated, which is even more rare (probably once per day). With that little usage, the total CPU % going to hkdf() will be negligible.